Missing 2m height coordinate and monotonicity for tasmin in CESM2 and CESM2-WACCM

Karen-A-Garcia opened the pull request:
This is a follow up to issue #2573 for the height coordinates. Still looking into the time issue. There was already a fix for the 2m height previously done for tas and I simply applied the fix to tasmin in both CESM2 and CESM2-WACCM files. There were already tests written for the fixes so I didn't add anything to those files.

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [ ] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

Karen-A-Garcia commented:
Thanks @valeriupredoi and thanks for showing me a similar fix for the monotonicity too. I've added a fix for the monotonicity as well as adding tests for them and I ran pre-commit before sending pushing the changes. Let me know if there's anything else I need to change.

valeriupredoi commented on esmvalcore/cmor/_fixes/cmip6/cesm2.py:
```diff
@@ -286,3 +287,80 @@ def fix_metadata(self, cubes):
         return cubes
+
+
+class Tasmin(Tas):
```
```suggestion
class Tasmin(Fix):
```

valeriupredoi replied:
I think the base class should be `Fix` here

valeriupredoi commented on tests/integration/cmor/_fixes/cmip6/test_cesm2.py:
```diff
@@ -420,3 +421,74 @@ def test_fgco2_fix_metadata():
+    def test_get(self):
+        """Test fix get."""
+        self.assertListEqual(
+            Fix.get_fixes("CMIP6", "NCAR", "day", "tasmin"),
```
```suggestion
            Fix.get_fixes("CMIP6", "CESM2", "day", "tasmin"),
```

valeriupredoi commented on tests/integration/cmor/_fixes/cmip6/test_cesm2.py:
```diff
@@ -420,3 +421,74 @@ def test_fgco2_fix_metadata():
+        """Test fix get."""
+        self.assertListEqual(
+            Fix.get_fixes("CMIP6", "NCAR", "day", "tasmin"),
+            [Tasmin(None)],
```
```suggestion
            [Tasmin(None), GenericFix(None)],
```

valeriupredoi commented on tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py:
```diff
@@ -124,3 +125,74 @@ def test_get_tas_fix():
+    def test_get(self):
+        """Test fix get."""
+        self.assertListEqual(
+            Fix.get_fixes("CMIP6", "NCAR", "day", "tasmin"),
```
```suggestion
            Fix.get_fixes("CMIP6", "CESM2-WACCM", "day", "tasmin"),
```

valeriupredoi commented on tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py:
```diff
@@ -124,3 +125,74 @@ def test_get_tas_fix():
+        """Test fix get."""
+        self.assertListEqual(
+            Fix.get_fixes("CMIP6", "NCAR", "day", "tasmin"),
+            [Tasmin(None)],
```
```suggestion
            [Tasmin(None), GenericFix(None)],
```

valeriupredoi reviewed (changes requested):
many thanks @Karen-A-Garcia :beer: A few wiggles to fix (am not 100% sure of the baseclass for Tasmin for CESM2, I *think* it should be a generic `Fix` class there, doesn't really matter that much). One general comment related to testing: it's great that you added tests, this is always a good thing to see from users, but wrt the testing framework, I'd suggest using the Pytest test classes/functions, and not use Unittest test classes anymore - these are old school and a tad harder to adapt inside a Pytest testing framework, plus, they need a lot more indentation :grin:

Karen-A-Garcia commented:
Thank you @valeriupredoi for looking over my code.🍺 I've changed all the old test formats and changed them so that they run with pytest. I had no issues on my end with the tests when I ran them in my terminal but let me know if there's anything else I need to change. Also looks like CESM2-FV2 and CESM2-WACCM-FV2 had some monotonicity issue so I've import changes to Tas and Pr made in CESM2 into those files.

valeriupredoi commented on tests/integration/cmor/_fixes/cmip6/test_cesm2_waccm.py:
```diff
@@ -124,3 +125,74 @@ def test_get_tas_fix():
+    def test_get(self):
+        """Test fix get."""
+        self.assertListEqual(
+            Fix.get_fixes("CMIP6", "NCAR", "day", "tasmin"),
```
hi @Karen-A-Garcia - these tests have not been corrected: they are still in UnitTest format, but a bit more critical, are not testing for the correct model, and also have issues with the actual tested outcome vs expected, could you please have another look at them? My code suggestion change appears to have been resolved, but the model is still wrong: you should test for CESM2-WACCM, not NCAR. Perhaps you forgot to push the changes? :beer:

Karen-A-Garcia replied:
Oh that is so strange because it was changed on my files when I pushed them and they seemed to not have change for some reason. I'll try to fix that.

valeriupredoi replied:
No worries at all! I was suspecting it's a push issue. If it's too much trouble to convert the tests to simple pytest funcs, don't worry about it, it's great you plopped tests in anyways. Let me know if I can help in any way 👍🍺

valeriupredoi commented:
many thanks @Karen-A-Garcia ! There are still some issues, unfortunately, please see the failed tests, and my comment above, perhaps you forgot to push to remote? Cheers! PS: I turned on Github Actions tests, since it's a lot easier to see test fails there than on CircleCI, see eg [test run output](https://github.com/ESMValGroup/ESMValCore/actions/runs/12072820422/job/33667632454?pr=2574)

Karen-A-Garcia commented:
@valeriupredoi Okay I think I've changed all the UnitTest formats into pytest now and the failed tests are coming from that code that I was previously talking about that was failing when I ran it in my terminal as well.

valeriupredoi commented:
excellent work @Karen-A-Garcia - many thanks! All spiffy now: the issue was that the `Tas()` fix was returning both the original `cubes` and the newly fixed cubes in the `new_cubes` iris `CubeList()`, and that the test was comparing a Python list with an iris `CubeList()` construct. Fixed that and it's ready to be merged when all tests pass :beer:

valeriupredoi reviewed (approved):
very many thanks @Karen-A-Garcia :beers:
